File: nikkake_flutter/lib/screens/home_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../constants/colors.dart';
import '../domain/date_utils.dart';
import '../domain/stats.dart';
import '../models/models.dart';
import '../providers/app_state.dart';
import '../widgets/ui.dart';
import 'routine_form_screen.dart';
import 'workout_screen.dart';

/// ホーム。アプリを開いた直後に見る画面。
/// サインインの有無に関係なく、ここに「今日やるルーティン」が並んでいて
/// 1タップで開始できる状態にしておくことがこの画面の唯一の役目。
class HomeScreen extends StatelessWidget {
  const HomeScreen({super.key});

  @override
  Widget build(BuildContext context) {
    final state = context.watch<AppState>();
    final streak = calculateStreak(state.routineLogs);

    final today = state.todayRoutines;
    final due = today.where((t) => t.isDueToday && !t.isCompleted).toList();
    final later = today.where((t) => !t.isDueToday && !t.isCompleted).toList();
    final done = today.where((t) => t.isCompleted).toList();

    return RefreshIndicator(
      onRefresh: () async => state.refresh(),
      child: ListView(
        key: const Key('home-screen'),
        padding: const EdgeInsets.all(Spacing.md),
        children: [
          _Header(),
          const SizedBox(height: Spacing.lg),
          _StreakBanner(streak: streak),
          const SizedBox(height: Spacing.xl),
          const Text(
            '今日のルーティン',
            style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold, color: AppColors.darkTextPrimary),
          ),
          const SizedBox(height: Spacing.md),

          if (today.isEmpty)
            EmptyState(
              key: const Key('home-empty'),
              icon: '📋',
              title: 'ルーティンがありません',
              message: 'まずは1つ作ってみましょう。作ったその日から記録が残ります。',
              action: AppButton(
                label: 'ルーティンを作る',
                onPressed: () => Navigator.of(context).push(
                  MaterialPageRoute(builder: (_) => const RoutineFormScreen()),
                ),
              ),
            )
          else if (due.isEmpty && later.isEmpty)
            const EmptyState(
              key: Key('home-all-done'),
              icon: '🎉',
              title: '今日の分は完了です',
              message: 'おつかれさま。この調子で続けましょう。',
            ),

          for (final item in due) _RoutineCard(item: item, state: _CardState.due),

          if (later.isNotEmpty) ...[
            const _SubTitle('今日は予定なし'),
            for (final item in later) _RoutineCard(item: item, state: _CardState.later),
          ],

          if (done.isNotEmpty) ...[
            const _SubTitle('完了'),
            for (final item in done) _RoutineCard(item: item, state: _CardState.done),
          ],
        ],
      ),
    );
  }
}

class _Header extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final now = DateTime.now();
    const weekdays = ['月', '火', '水', '木', '金', '土', '日'];

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          greetingForHour(now.hour),
          style: const TextStyle(fontSize: 28, fontWeight: FontWeight.bold, color: AppColors.darkTextPrimary),
        ),
        const SizedBox(height: Spacing.xs),
        Text(
          '${now.month}月${now.day}日 (${weekdays[now.weekday - 1]})',
          style: const TextStyle(fontSize: 14, color: AppColors.darkTextSecondary),
        ),
      ],
    );
  }
}

class _StreakBanner extends StatelessWidget {
  final StreakInfo streak;

  const _StreakBanner({required this.streak});

  @override
  Widget build(BuildContext context) {
    return AppCard(
      key: const Key('streak-banner'),
      color: AppColors.darkSurfaceElevated,
      child: Row(
        children: [
          Text(streak.current > 0 ? '🔥' : '🌱', style: const TextStyle(fontSize: 32)),
          const SizedBox(width: Spacing.md),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                const Text('連続記録', style: TextStyle(fontSize: 12, color: AppColors.darkTextSecondary)),
                Text(
                  '${streak.current} 日',
                  key: const Key('streak-count'),
                  style: const TextStyle(
                    fontSize: 20,
                    fontWeight: FontWeight.bold,
                    color: AppColors.darkAccent,
                  ),
                ),
              ],
            ),
          ),
          Column(
            crossAxisAlignment: CrossAxisAlignment.end,
            children: [
              const Text('最長', style: TextStyle(fontSize: 12, color: AppColors.darkTextSecondary)),
              Text(
                '${streak.longest} 日',
                style: const TextStyle(
                  fontSize: 16,
                  fontWeight: FontWeight.bold,
                  color: AppColors.darkTextPrimary,
                ),
              ),
            ],
          ),
        ],
      ),
    );
  }
}

class _SubTitle extends StatelessWidget {
  final String text;

  const _SubTitle(this.text);

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.only(top: Spacing.md, bottom: Spacing.sm),
      child: Text(
        text,
        style: const TextStyle(fontSize: 13, fontWeight: FontWeight.bold, color: AppColors.darkTextSecondary),
      ),
    );
  }
}

enum _CardState { due, later, done }

class _RoutineCard extends StatelessWidget {
  final TodayRoutine item;
  final _CardState state;

  const _RoutineCard({required this.item, required this.state});

  @override
  Widget build(BuildContext context) {
    final routine = item.routine.routine;

    return Padding(
      padding: const EdgeInsets.only(bottom: Spacing.sm),
      child: Opacity(
        opacity: state == _CardState.due ? 1 : 0.55,
        child: InkWell(
          key: Key('routine-card-${routine.id}'),
          borderRadius: BorderRadius.circular(Radii.lg),
          onTap: () => Navigator.of(context).push(
            MaterialPageRoute(builder: (_) => WorkoutScreen(routineId: routine.id)),
          ),
          child: AppCard(
            child: Row(
              children: [
                Container(
                  width: 48,
                  height: 48,
                  decoration: BoxDecoration(
                    color: hexToColor(routine.color),
                    shape: BoxShape.circle,
                  ),
                  alignment: Alignment.center,
                  child: Text(routine.icon, style: const TextStyle(fontSize: 24)),
                ),
                const SizedBox(width: Spacing.md),
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(
                        routine.name,
                        style: const TextStyle(
                          fontSize: 16,
                          fontWeight: FontWeight.bold,
                          color: AppColors.darkTextPrimary,
                        ),
                      ),
                      const SizedBox(height: 2),
                      Text(
                        state == _CardState.done
                            ? '今日は完了しました'
                            : '${formatFrequency(routine)} ・ ${item.routine.exercises.length}種目',
                        style: const TextStyle(fontSize: 12, color: AppColors.darkTextSecondary),
                      ),
                    ],
                  ),
                ),
                Text(
                  state == _CardState.done ? '✅' : '▶',
                  style: TextStyle(
                    fontSize: 18,
                    color: state == _CardState.done ? null : AppColors.darkPrimaryLight,
                  ),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

File: nikkake_kmp/shared/src/commonMain/kotlin/com/myapplication/common/ui/screens/HomeScreen.kt
package com.myapplication.common.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.myapplication.common.data.StreakInfo
import com.myapplication.common.data.TodayRoutine
import com.myapplication.common.domain.calculateStreak
import com.myapplication.common.domain.currentHour
import com.myapplication.common.domain.formatFrequency
import com.myapplication.common.domain.greetingForHour
import com.myapplication.common.domain.today
import com.myapplication.common.store.AppStore
import com.myapplication.common.ui.Destination
import com.myapplication.common.ui.Navigation
import com.myapplication.common.ui.components.AppButton
import com.myapplication.common.ui.components.AppCard
import com.myapplication.common.ui.components.EmptyState
import com.myapplication.common.ui.components.tag
import com.myapplication.common.ui.theme.LocalPalette
import com.myapplication.common.ui.theme.Spacing
import com.myapplication.common.ui.theme.hexToColor

/**
 * ホーム。アプリを開いた直後に見る画面。
 * サインインの有無に関係なく、ここに「今日やるルーティン」が並んでいて
 * 1タップで開始できる状態にしておくことがこの画面の唯一の役目。
 */
@Composable
fun HomeScreen(appStore: AppStore, navigation: Navigation) {
    val palette = LocalPalette.current
    val streak = calculateStreak(appStore.routineLogs)

    val todayRoutines = appStore.todayRoutines
    val due = todayRoutines.filter { it.isDueToday && !it.isCompleted }
    val later = todayRoutines.filter { !it.isDueToday && !it.isCompleted }
    val done = todayRoutines.filter { it.isCompleted }

    LazyColumn(
        modifier = Modifier.fillMaxSize().background(palette.background).tag("home-screen"),
        contentPadding = PaddingValues(Spacing.md),
    ) {
        item {
            val date = today()
            Column {
                Text(
                    greetingForHour(currentHour()),
                    fontSize = 28.sp,
                    fontWeight = FontWeight.Bold,
                    color = palette.textPrimary,
                )
                Text(
                    "${date.monthNumber}月${date.dayOfMonth}日",
                    modifier = Modifier.padding(top = Spacing.xs),
                    fontSize = 14.sp,
                    color = palette.textSecondary,
                )
            }
            Spacer(Modifier.height(Spacing.lg))
            StreakBanner(streak)
            Spacer(Modifier.height(Spacing.xl))
            Text(
                "今日のルーティン",
                fontSize = 20.sp,
                fontWeight = FontWeight.Bold,
                color = palette.textPrimary,
            )
            Spacer(Modifier.height(Spacing.md))
        }

        if (todayRoutines.isEmpty()) {
            item {
                Box(Modifier.tag("home-empty")) {
                    EmptyState(
                        icon = "📋",
                        title = "ルーティンがありません",
                        message = "まずは1つ作ってみましょう。作ったその日から記録が残ります。",
                        action = {
                            AppButton("ルーティンを作る", onClick = {
                                navigation.push(Destination.RoutineForm())
                            })
                        },
                    )
                }
            }
        } else if (due.isEmpty() && later.isEmpty()) {
            item {
                Box(Modifier.tag("home-all-done")) {
                    EmptyState(
                        icon = "🎉",
                        title = "今日の分は完了です",
                        message = "おつかれさま。この調子で続けましょう。",
                    )
                }
            }
        }

        items(due, key = { it.routine.id }) { item ->
            RoutineCard(item, CardState.DUE) { navigation.push(Destination.Workout(item.routine.id)) }
        }

        if (later.isNotEmpty()) {
            item { SubTitle("今日は予定なし") }
            items(later, key = { it.routine.id }) { item ->
                RoutineCard(item, CardState.LATER) { navigation.push(Destination.Workout(item.routine.id)) }
            }
        }

        if (done.isNotEmpty()) {
            item { SubTitle("完了") }
            items(done, key = { it.routine.id }) { item ->
                RoutineCard(item, CardState.DONE) { navigation.push(Destination.Workout(item.routine.id)) }
            }
        }
    }
}

@Composable
private fun StreakBanner(streak: StreakInfo) {
    val palette = LocalPalette.current

    AppCard(modifier = Modifier.tag("streak-banner"), color = palette.surfaceElevated) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(if (streak.current > 0) "🔥" else "🌱", fontSize = 32.sp)
            Spacer(Modifier.size(Spacing.md))
            Column(Modifier.weight(1f)) {
                Text("連続記録", fontSize = 12.sp, color = palette.textSecondary)
                Text(
                    "${streak.current} 日",
                    modifier = Modifier.tag("streak-count"),
                    fontSize = 20.sp,
                    fontWeight = FontWeight.Bold,
                    color = palette.accent,
                )
            }
            Column(horizontalAlignment = Alignment.End) {
                Text("最長", fontSize = 12.sp, color = palette.textSecondary)
                Text(
                    "${streak.longest} 日",
                    fontSize = 16.sp,
                    fontWeight = FontWeight.Bold,
                    color = palette.textPrimary,
                )
            }
        }
    }
}

@Composable
private fun SubTitle(text: String) {
    Text(
        text,
        modifier = Modifier.padding(top = Spacing.md, bottom = Spacing.sm),
        fontSize = 13.sp,
        fontWeight = FontWeight.Bold,
        color = LocalPalette.current.textSecondary,
    )
}

private enum class CardState { DUE, LATER, DONE }

@Composable
private fun RoutineCard(item: TodayRoutine, state: CardState, onClick: () -> Unit) {
    val palette = LocalPalette.current
    val routine = item.routine.routine

    Box(
        modifier = Modifier
            .padding(bottom = Spacing.sm)
            .alpha(if (state == CardState.DUE) 1f else 0.55f)
            .tag("routine-card-${routine.id}")
            .clickable(onClick = onClick),
    ) {
        AppCard {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Box(
                    modifier = Modifier
                        .size(48.dp)
                        .background(hexToColor(routine.color), CircleShape),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(routine.icon, fontSize = 24.sp)
                }
                Spacer(Modifier.size(Spacing.md))
                Column(Modifier.weight(1f)) {
                    Text(
                        routine.name,
                        fontSize = 16.sp,
                        fontWeight = FontWeight.Bold,
                        color = palette.textPrimary,
                    )
                    Text(
                        if (state == CardState.DONE) "今日は完了しました"
                        else "${formatFrequency(routine)} ・ ${item.routine.exercises.size}種目",
                        fontSize = 12.sp,
                        color = palette.textSecondary,
                    )
                }
                Text(
                    if (state == CardState.DONE) "✅" else "▶",
                    fontSize = 18.sp,
                    color = if (state == CardState.DONE) palette.textPrimary else palette.primaryLight,
                )
            }
        }
    }
}
